File: sentimental-credit/credit.py
# get user input, digit length, convert to int, and set count to 0
c = input("Number: ")
n = len(c)
c = int(c)
count = 0

# calculate the "count" given by the algorithm
def count_func(c, n, count):

    # multiply every other digit by 2 starting with the 2nd to last digit and add to total count
    for i in [2, 4, 6, 8, 10, 12, 14, 16]:
        if (n == 16) or (n == 15 and i != 16) or (n == 13 and i != 16 and i != 14):
            temp = digit(c, i) * 2
            if temp < 10:
                count += temp
            else:
                count += round((temp/10) - ((temp % 10)/10)) + (temp % 10)

    # add remaining digits to the total count
    for i in [1, 3, 5, 7, 9, 11, 13, 15]:
        if (n == 16) or (n == 15) or (n == 13 and i != 15):
            temp = digit(c, i)
            count += temp

    return count

# return specific digit where nth is the nth digit starting from the end
def digit(c, nth):
    y = ((c % (10**nth)/(10**(nth - 1)))) - (c % (10**(nth - 1)) / (10**(nth - 1)))
    return round(y)

# print the result
def print_result(c, n):

    count_final = count_func(c, n, count)

    if (count_final % 10 != 0) or (n < 13):
        print("INVALID")
    elif n == 15 and digit(c, 15) == 3 and (digit(c, 14) == 4 or digit(c, 14) == 7):
        print("AMEX")
    elif (n == 16 and digit(c, 16) == 4) or n == 13:
        print("VISA")
    elif n == 16 and digit(c, 16) == 5 and digit(c, 15) < 6:
        print("MASTERCARD")
    else:
        print("INVALID")


print_result(c, n)

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    long n;
    long o;
    long p;
    int c;
    int q;
    int r;
    int s;
    int t;
    {
        n = get_long("Number: ");
        o = n;
        c = 0;
    }

    while (o > 10)
//calculate everyother digit x2 starting from 2nd to last
    {
        p = o % 10;
        o = (o - p)/10;
        p = o % 10;
        q = (p * 2);
        if (q > 9)
        {
            r = q % 10;
            s = (q - r)/10;
            c = c + r + s;
        }
        else
        {
            c = c + (p * 2);
        }
        p = o % 10;
        o = (o - p)/10;
        p = o % 10;
    }
    o = n;
    while (o > 1)
//calculate the sum of remaining digits
    {
        p = o % 10;
        c = c + p;
        o = (o - p)/10;
        p = o % 10;
        o = (o - p)/10;
    }
//check if ends in 0
    t = c % 10;
    if (t != 0)
    {
        printf("INVALID\n");
    }
    else
    {
// check for 13 digit VISA
        o = n % 1000000000000;
        p = (n - o)/1000000000000;
        if (p < 10){
            if (p == 4){
            printf("VISA\n");
            }
            else{
                printf("INVALID\n");
            }
        }
//check for 15 digit AMEX
        else if (p > 9 && p < 999){
            o = p % 10;
            p = (p - o) / 10;
            if (p == 34 || p == 37){
                printf("AMEX\n");
                }
            else{
                printf("INVALID\n");
            }
        }
 //check if 16 digit then check if Mastercard, Visa, or Invalid
            else if (p > 999){
                o = p % 100;
                p = (p - o) / 100;
                if (p == 51 || p == 52 || p == 53 || p == 54 || p == 55){
                    printf("MASTERCARD\n");
                }
                else if (p > 9){
                    o = p % 10;
                    p = (p - o) / 10;
                    if (p == 4){
                        printf("VISA\n");
                }
                    else{
                    printf("INVALID\n");
                }
                }

            }

            }

            }